### ISSUE TITLE
[Feature request] Auto-restart failed job

### ISSUE THREAD

**Issue by mario45211** [hazelcast/hazelcast-jet#3041]
Hello Jet Team! :)

I'm developing a stateless stream pipeline under Jet engine. One of my requirements is its continuous working, regardless of errors from any operation within it. 

From the Jet's documentation I read: _One option for dealing with failure in immutable-state jobs is simply restarting them_, but have you ever consider adding auto-restarting option ? Once my job fails I must restart it manually, which is highly undesirable. 

Of course, there is a workaround to implement external watcher to monitor and do restart job if failed one was detected, but mantaining and deploying additional watcher is cumbersome, especially when job is deployed to Jet's cluster. 

Please, leave below some comment what you think about such feature.

Best regards

**Comment by vladoschreiner** [comment]
Hi Mario,

**Comment by vladoschreiner** [comment]
Hi Mario, what kind of failures cause your job to fail? If you configure the sources/sinks to retry infinitely and you cover exceptions in the user-provided code, the engine should keep the job running.

**Comment by mario45211** [comment]
Unfortunately, I do not have any revelant example, but from my experience there is some type of errors which cause a job to fail, where try-catch blocks on my code are not enough.

I'll attach some example as soon as I catch failure from any execution.

**Comment by vladoschreiner** [comment]
Thanks. The engine shall recover from recoverable errors (lost cluster member, lost source/sink connectivity) and give you the tools to deal with issues outside of its control (data outliers, exceptions in your code). Issues "beyond Jet " (JVM crashes, provision HW if it fails) are deliberately delegated to other SW (K8s or so). That said, there shouldn't be an issue where a blind job auto-restart helps. So let's discuss case by case.

**Comment by hhromic** [comment]
Hi @vladoschreiner , in our team we are facing the exact same situation with a Kafka-to-JDBC pipeline.
In one instance, the database itself went down for a long period of time (hours) and our Jet pipeline failed out after a number of re-connection retries. We had to manually restart the pipeline job to resume it.

Can I ask you how exactly you configure infinite retries as you suggested in your comment?
I am unable to find anything about it in the Jet docs, apidocs nor the source code. Thanks!

**Comment by viliam-durina** [comment]
Specifically in the case of the JDBC sink, we retry the connection if the exception we got from the JDBC driver wasn't a `SQLNonTransientException`. Maybe in your case the driver threw some other exception, but after something changed it threw a `SQLNonTransientException` after which the sink stopped retrying and the job failed.

We are indeed considering adding an auto-restart feature. A streaming job should never completely fail, it's supposed to run until the user cancels it. Issue #1126 is also related.

**Comment by hhromic** [comment]
You are right, the JDBC driver (Vertica) is indeed throwing non-transient exceptions here.

FIrst, when the Vertica cluster went down:
```
23:43:42.992 [ WARN] [c.h.j.i.c.WriteJdbcP] [(redacted)/jdbcSink#1] Exception when connecting and preparing the statement,
java.sql.SQLNonTransientConnectionException: [Vertica][VJDBC](100176) Failed to connect to host (redacted) on port 5433. Reason: Failed to establish a connection to the primary server or any backup address.,
(...)
Caused by: java.io.IOException: Failed to establish a connection to the primary server or any backup address.,
(...)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
```

Then when the cluster came back but was not yet ready:
```
23:44:47.645 [ WARN] [c.h.j.i.c.WriteJdbcP] [(redacted)/jdbcSink#1] Exception when connecting and preparing the statement,
java.sql.SQLNonTransientConnectionException: [Vertica][VJDBC](4149) FATAL: Node startup/recovery in progress. Not yet ready to accept connections
```

And finally when Jet gave up:
```
23:46:17.788 [ERROR] [c.h.j.i.MasterJobContext] Execution of job '(redacted)', execution 05b9-de07-20c1-0017 failed,
	Start time: 2021-02-12T13:45:51.214,
	Duration: 4d 10:00:26.573,
com.hazelcast.jet.JetException: Exception in ProcessorTasklet{(redacted)/jdbcSink#0}: com.hazelcast.jet.JetException: java.sql.SQLNonTransientException: [Vertica][VJDBC](4971) ERROR: The status of one or more nodes changed during query planning
```

This means it is not really Jet's fault. If these exceptions should be transient or non-transient is subject to debate and I doubt it will be changed by upstream in any foreseable future :(

Note that we did not touch the Jet cluster during the above sequence of exceptions (which happened over several hours until the job got cancelled). So Jet did retry the sink connection for a while until it gave up and did not instantly cancel the job on first exception.

All that being said, adding an auto-restart feature in Jet would be a great addition to the engine for us so we are highly interested on seeing this in the future for Jet 👍

**Comment by viliam-durina** [comment]
I'd say we should not give up trying even in case of a non-transient exception. The boundary between transient and non-transient is very unclear - TCP connection refused is transient, if the target is down, but is not, if the host name is wrong. DB constraint violation is transient if somebody created a stricter constraint by mistake and is going to delete it, but is non-transient if there's a data error. And so on.

What we plan to do is to treat all exceptions in all processors as transient, if the job is fault-tolerant. Or if enabled by the user explicitly.

**Comment by vladoschreiner** [comment]
> If these exceptions should be transient or non-transient is subject to debate and I doubt it will be changed by upstream in any foreseable future

That is a fair point: if we rely just on the underlying (JDBC) driver to decide what is transient (recoverable) and what is not we can't really guarantee the behaviour for Jet users. And we can't of course test all possible driver implementations.

Thus, having a single "keep trying" option will help operating the job in shaky conditions. 

We are adding this problem to the Jet roadmap. Please contribute with other relevant findings and bug reports.

**Comment by hhromic** [comment]
Hi @viliam-durina @vladoschreiner, we are still seeing this issue in our setups :(
We are wondering if you still have plans to address the problem or if you removed this from your roadmap?
We are migrating our Jet-based applications using 4.x to the new Hazelcast Platform 5.x version.
Perhaps this issue should be moved to the new consolidated repository so it doesn't get lost/forgotten? :)

Thanks again for Hazelcast (Jet). It is truly a great stream processing platform.
